vespalib/src/vespa/vespalib/stllike/hashtable.h

```diff
@@ -141,39 +141,38 @@ class hashtable : public hashtable_base
         typedef Value* pointer;
         typedef std::forward_iterator_tag iterator_category;
 
-        iterator(hashtable * hash, next_t start) : _hash(start), _subNode(start), _hashTable(hash) {
-            advanceToNextValidHash();
-        }
-        iterator(hashtable * hash, next_t start, next_t subNode) : _hash(start), _subNode(subNode), _hashTable(hash) { }
-        Value & operator * ()  const { return _hashTable->get(_subNode); }
-        Value * operator -> () const { return & _hashTable->get(_subNode); }
-        iterator & operator ++ () {
-            if (_hashTable->_nodes[_subNode].hasNext()) {
-                _subNode = _hashTable->_nodes[_subNode].getNext();
-            } else {
-                _hash++;
+        iterator(hashtable * hash) : _current(0), _hashTable(hash) {
+            if ((_current < _hashTable->initializedSize()) && ! _hashTable->_nodes[_current].valid()) {
                 advanceToNextValidHash();
             }
+        }
+        iterator(hashtable * hash, next_t pos) : _current(pos), _hashTable(hash) { }
+        static iterator end(hashtable *hash) { return iterator(hash, Node::npos); }
+
+        Value & operator * ()  const { return _hashTable->get(_current); }
+        Value * operator -> () const { return & _hashTable->get(_current); }
+        iterator & operator ++ () {
+            advanceToNextValidHash();
             return *this;
         }
         iterator operator ++ (int) {
             iterator prev = *this;
             ++(*this);
             return prev;
         }
-        bool operator==(const iterator& rhs) const { return (_subNode == rhs._subNode); }
-        bool operator!=(const iterator& rhs) const { return (_subNode != rhs._subNode); }
+        bool operator==(const iterator& rhs) const { return (_current == rhs._current); }
```

Comment:
Consider adding an assert on both iterators coming from the same hash table for == and !=

Reply:
I think that is too conservative. I think we expect this much form users of iterators.